### PR TITLE
[PINOT-7088] No-downtime table rebalance

### PR DESCRIPTION
**Pull request by sunithabeeram**
Updated logic in #3210. 
Currently, when the controller's rebalance API is called, the idealstate is updated in one go and there are no guarantees around replica availability.

This change provides a way to do no downtime rebalancing. Key changes include:

1. Option to specify via rebalance API whether a downtime is acceptable or not. (By default the code does the safe thing: no-downtime rebalance).
2. Provide a helper class to handle the actual ideal-state update actions. This is moved out of the Strategy classes - perhaps a clearer approach as the Strategy classes just inform what needs to be done and the actual execution is handled outside of the Strategy.
3. IdealStateUpdater class: This class does the no-downtime update using a fixed thread pool - this allows for the API to return the target IdealState to the user and handle the execution in the background. The no downtime aspect is handled thus:
- The ideal state is updated to reflect the target state first ensuring there is atleast one serving replica. For ex, if current state is Segment 1 -> {host1:online, host2:online} and new state is Segment 1 -> {host3:online, host4:online}, we transition through the following changes:
Segment 1 -> {host1:online, host3:online}
Similar transitions for segments that don't have any common replica for the target state are triggered.
- After this the ideal state is set to the target which will result in removing other replicas from the segments: Segment1 -> {host3:online, host4:online}

Testing done:
- Added basic coverage to ensure the one-replica guarantee is maintained.
- TODO: add tests to cover more of the added code.

Manual testing at this point to ensure:
- API return immediately
- Table with ReplicaGroupAssignmentStrategy can be rebalanced correctly when instances per partition are changed (ie, added or removed).
- TODO: Test end to end with realtime segments and DefaultSegmentAssignmentStrategy.
- Manual testing uncovered a problem with partition assignment that is tracked in: https://github.com/linkedin/pinot/issues/3216.

### CODE REVIEW

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/IdealStateUpdater.java [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+
+    _executorService = Executors.newFixedThreadPool(MAX_THREADS,
+        new ThreadFactoryBuilder().setNameFormat("idealstate-updater-thread-%d").build());
+  }
+
+  /**
+   * Updates idealstate to be rebalanced. The update is handled in a background thread to track the progress of the
+   * update.
+   */
+  public void update(IdealState targetState, TableConfig config, RebalanceSegmentStrategy strategy,
```
[the remark this answers is not in the document]
Comment (sunithabeeram) [reply pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/IdealStateUpdater.java]: I went back and forth a bit on this - I initially had it not taking a targetState; but the API expects the ideal state to be returned in response. for both dry-run and non-dry run cases. It felt weird to call the updater when it was in dry run mode, so I had the manager compute it once and updater will recompute it only if needed. I see your other comment on the name of this class being misleading. Let me rework this a bit to have things flow cleanly.

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/IdealStateUpdater.java [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+        }
+        // in case of any error, we retry - currently an unbounded number of times
+      } catch (ZkBadVersionException e) {
+        LOGGER.warn("Version changed while updating ideal state for resource: {}", tableName);
+      } catch (Exception e) {
+        LOGGER.warn("Caught exception while updating ideal state for resource: {}", tableName, e);
+      }
+
+      previousIdealState = currentIdealState;
+      // wait before retrying
```
[the remark this answers is not in the document]
Comment (sunithabeeram) [reply pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/IdealStateUpdater.java]: We are using a fixed delay retry so as to not hit zk too hard.

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/IdealStateUpdater.java [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+        if (zkBaseDataAccessor.set(idealStateKey.getPath(), nextIdealState.getRecord(),
+            currentIdealState.getRecord().getVersion(), AccessOption.PERSISTENT)) {
+          // if we succeeded, wait for the change to stabilize
+          waitForStable(tableName);
+          continue;
+        }
+        // in case of any error, we retry - currently an unbounded number of times
+      } catch (ZkBadVersionException e) {
+        LOGGER.warn("Version changed while updating ideal state for resource: {}", tableName);
+      } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (sunithabeeram) [reply pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/IdealStateUpdater.java]: The other exceptions I can anticipate are network errors which can be retried. Looking through ZkBaseDataAccessor, set() returns false if there was any other exception internally. It is unclear which ones can be retried. I will bound it to a fixed number of times per idealstate update attempt.

File: pinot-common/src/main/java/com/linkedin/pinot/common/utils/helix/HelixHelper.java [hunk 4]
```diff
@@ -56,6 +56,12 @@
 
   public static final String BROKER_RESOURCE = CommonConstants.Helix.BROKER_RESOURCE_INSTANCE;
 
+  public static IdealState cloneIdealState(IdealState idealState) {
+    ZNRecordSerializer znRecordSerializer = new ZNRecordSerializer();
```
Comment (kishoreg) [inline pinot-common/src/main/java/com/linkedin/pinot/common/utils/helix/HelixHelper.java]: nit: check if we can use have a static instance of znRecordSerializer

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java [hunk 5]
```diff
@@ -2084,18 +2084,20 @@ public JSONObject rebalanceTable(final String rawTableName, TableType tableType,
...
+        _tableRebalancer.rebalance(tableConfig, rebalanceSegmentsStrategy, rebalanceUserConfig);
+        jsonObject.put("status", "Rebalance triggered. Check logs for progress");
+      } else {
+        // dry run mode: return the target partition assignment and idealstate mapping
+        PartitionAssignment newPartitionAssignment =
+            rebalanceSegmentsStrategy.rebalancePartitionAssignment(idealState, tableConfig, rebalanceUserConfig);
+        IdealState targetState =
+            rebalanceSegmentsStrategy.getRebalancedIdealState(idealState, tableConfig, rebalanceUserConfig);
+        jsonObject.put("partitionAssignment", newPartitionAssignment.getPartitionToInstances());
+        jsonObject.put("idealState", targetState.getRecord().getMapFields());
```
Comment (kishoreg) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java]: not part of this PR, but this is method is mixing the responsibility of controllerRestResource and PinotHelixResourceManager. Returning a json object from this method is not a good idea. Also, this should be sync and we can make it async from the RestApi if needed.

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java [hunk 6]
```diff
@@ -2084,18 +2084,20 @@ public JSONObject rebalanceTable(final String rawTableName, TableType tableType,
...
-
-      // update it as needed
+      // rebalance it as needed
       if (!rebalanceUserConfig.getBoolean(RebalanceUserConfigConstants.DRYRUN)) {
         // call the updater providing it a reference to the strategy so it can recompute target state if needed
-        _idealStateUpdater.update(targetState, tableConfig, rebalanceSegmentsStrategy, rebalanceUserConfig);
+        _tableRebalancer.rebalance(tableConfig, rebalanceSegmentsStrategy, rebalanceUserConfig);
+        jsonObject.put("status", "Rebalance triggered. Check logs for progress");
+      } else {
+        // dry run mode: return the target partition assignment and idealstate mapping
```
Comment (kishoreg) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java]: can we do this part of tableRebalancer itself?  rebalance method should return the final idealstate.

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/IdealStateUpdater.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (kishoreg) [reply pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/IdealStateUpdater.java]: Good idea, I am assuming this will happen only for unknown exceptions.

File: pinot-controller/src/test/java/com/linkedin/pinot/controller/helix/core/util/TableRebalancerTest.java [hunk 7]
```diff
@@ -72,22 +72,22 @@ public void downtimeUpdateWithCommonElements() {
...
     updater.updateSegmentIfNeeded(segmentId, srcMap, targetMap, current, downtime);
 
     Map<String, String> tempMap = current.getRecord().getMapField(segmentId);
     Assert.assertTrue(EqualityUtils.isEqual(tempMap, targetMap));
   }
 
-  // no-downtime update without common elements - target state is updated to have one up replica
+  // no-downtime rebalance without common elements - target state is updated to have one up replica
   @Test
   public void NoDowntimeUpdateWithNoCommonElements() {
```
Comment (kishoreg) [inline pinot-controller/src/test/java/com/linkedin/pinot/controller/helix/core/util/TableRebalancerTest.java]: No -> no
